pkg/controller/drain/drain_controller.go

```diff
@@ -0,0 +1,417 @@
+	defer func() {
+		glog.V(4).Infof("Finished syncing node %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// First check if a drain request is needed for this node
+	// TODO (jerzhang): consider not using the cache for this,
+	// since it has been shown to be problematic in the past
+	node, err := ctrl.nodeLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("node %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	desiredState := node.Annotations[daemonconsts.DesiredDrainerAnnotationKey]
+	if desiredState == node.Annotations[daemonconsts.LastAppliedDrainerAnnotationKey] {
+		glog.V(4).Infof("Node %v has the correct drain", key)
+		return nil
+	}
+
+	podName, ok := os.LookupEnv("POD_NAME")
+	if !ok || podName == "" {
+		glog.Fatalf("pod name is required for controller drainer")
+	}
+
+	drainer := &drain.Helper{
+		Client:              ctrl.kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteEmptyDirData:  true,
+		GracePeriodSeconds:  -1,
+		Timeout:             90 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			glog.Infof("%s pod %s/%s", verbStr, pod.Namespace, pod.Name)
+		},
+		AdditionalFilters: []drain.PodFilter{
+			func(pod corev1.Pod) drain.PodDeleteStatus {
+				if pod.Name == podName {
+					return drain.MakePodDeleteStatusSkip()
+				}
+				return drain.MakePodDeleteStatusOkay()
+			},
+		},
+		Out:    writer{glog.Info},
+		ErrOut: writer{glog.Error},
+		Ctx:    context.TODO(),
+	}
+
+	desiredVerb := strings.Split(desiredState, "-")[0]
+	if desiredVerb == daemonconsts.DrainerStateUncordon {
+		glog.Infof("node %v to be uncordoned", key)
+		// perform uncordon
+		if err := ctrl.cordonOrUncordonNode(false, node, drainer); err != nil {
+			return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
+		}
+	}
+
+	if desiredVerb == daemonconsts.DrainerStateDrain {
+		// First check if we have an ongoing drain
+		// This is currently stored in the object itself as a map but,
+		// Practically during upgrades the control plane node this controller
+		// pod is running on will also be terminated (the drainer will skip it).
+		// This is a bit problematic in practice since we don't really have a previous state.
+		// TODO (jerzhang) consider using a new CRD for coordination
+
+		glog.Infof("node %v to be drained", key)
+		ongoingDrain := false
+		for k, v := range ctrl.ongoingDrains {
+			if k == node.Name {
+				glog.Infof("TEST: previous node drain found. Node %v time %v", k, v)
+				ongoingDrain = true
+				duration := time.Now().Sub(v).Hours()
+				glog.Infof("Drain has been going on for %v hours", duration)
+				if duration > 1 {
+					return fmt.Errorf("Drain has exceeded the 1 hour error mark")
```

Comment:
Then avoid repeating it with e.g.:
`return fmt.Errorf("Drain exceeded timeout: %v", drainMaxDuration)`
or so